src/ucp/api/ucp.h

```diff
@@ -1651,9 +1670,56 @@ typedef struct ucp_mem_map_params {
       * set to NULL by default.
       */
     const void              *exported_memh_buffer;
+
+    /**
+     * Mkey index to use when doing memory registration.
+     * This field is optional.
+     *
+     * This parameter is application defined and it is used to produce memory
+     * handles that will pack into predictable rkeys. With that capability, it
+     * becomes possible to avoid exchanging rkeys, as the remote endpoint can
+     * directly build them, provided that it uses the same mkey index.
```

Comment:
```suggestion
     * This parameter is application-defined and is used to produce memory
     * handles for which rkeys can be computed remotely without requiring 
     * any data exchange with the owner of the memory region.
     * Knowing the logical Mkey index, the remote endpoint will be able
     * to directly (locally) build the corresponding rkey.
```

Reply:
is it going to be supported on all system types?
what happens if the underlying layer does not support it?

Reply:
Good point!
I think that the target system should get ‘ERR_NOT_SUPPORTED’ when doing ucp_mem_map.
The initiator side should probably negotiate with the target if the logical mkey indexing is supported.

Reply:
Also it is possible that while logical mkey indexing is supported, the app will run out of resource.

At the device level we have a way to query the amount of resources available. I think we may want to expose that on the target side in a form of the number of available logical mkey indexes. In UCP it’s complicated be the fact that multiple transports are available and some may support this and some not.
I’m not sure what’s the best way to communicate that. Maybe query worker for the inter-node-only capabilities (for which this feature is mainly useful) to see how many logical mkey indexes the networking transport provides.

Reply:
I don't like the "unsupported" approached because it is going against core design principles of UCX - write code once and it works everywhere.  This is where this API proposal breaks apart IMHO, because it changes the entire workflow of the application, like exchanging RKEYS, etc.

Reply:
I don't think it changes that. At the rkey exchange step you'd exchange with other peers the fact that rkey's can be generated locally and no exchange is needed